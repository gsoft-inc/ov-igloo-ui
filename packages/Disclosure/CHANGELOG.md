# @igloo-ui/disclosure

## 1.1.0

### Minor Changes

- 4846c59: Wrapped Igloo CSS in @layer to isolate it.

## 1.0.4

### Patch Changes

- cfb5619: Updated hopper token versions and button versions.

## 1.0.3

### Patch Changes

- 3944b5f: Fixed Disclosure styles including font sizes, aligment and focus styles.

## 1.0.2

### Patch Changes

- 7d36815: Fixed the disclosure icon alignment.

## 1.0.1

### Patch Changes

- a2df55d: Updated to latest Hopper Token dependency

## 1.0.0

### Major Changes

- c981a8b: Separated the controlled and uncontrolled expanded props.

  ## BREAKING CHANGE

  The defaultExpanded prop was added, which means isExpanded may need to be exchanged for this new prop.
  If you use isExpanded to make the disclosure expanded by default, rename it to defaultExpanded.
  If you use isExpanded to open/close the disclosure programmatically, keep isExpanded.
  Rule of thumb: If isExpanded was used without onOpen and onClose, it most likely has to be renamed to defaultExpanded.

## 0.3.1

### Patch Changes

- 9739155: Gave ColorPicker the new Wokleap look. Color now uses css variables to update its color. Fixed spacing issues in tooltip. Updated hopper token version.

## 0.3.0

### Minor Changes

- a66541c: Gave the Disclosure component the new Brand

## 0.2.3

### Patch Changes

- 08de993: Updated react-aria packages

## 0.2.2

### Patch Changes

- bf4ddef: Updated storybook version and linting. Made linting fixes in these components.

## 0.2.1

### Patch Changes

- c3efaf5: Fixed tests and snapshots since react-aria generated ids changed and viewBox was added to icons.

## 0.2.0

### Minor Changes

- 210d1a8: Added isLowContrast, onClose and onOpen props. isExpended will now update the open and close state whenever updated.

## 0.1.1

### Patch Changes

- a41e1d4: Fix: moved destructuring of props to function parameters so that storybook will automatically show the default value in the props table. It was done in Modal because it needed to send props to useDialog. In this case, @default annotation was used in comments.

## 0.1.0

### Minor Changes

- d6f1db8: Initial release of disclosure component
