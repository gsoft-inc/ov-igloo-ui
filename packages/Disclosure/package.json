{
    "name": "@igloo-ui/disclosure",
    "version": "1.2.3",
    "main": "dist/Disclosure.js",
    "module": "dist/Disclosure.js",
    "types": "dist/Disclosure.d.ts",
    "styles": "dist/disclosure.css",
    "publishConfig": {
        "access": "public"
    },
    "repository": "git+https://github.com/gsoft-inc/ov-igloo-ui.git",
    "author": "Officevibe Inc.",
    "contributors": [
        "Vicky Comeau vicky.comeau@gsoft.com"
    ],
    "license": "Apache-2.0",
    "peerDependencies": {
        "react": ">= 16.8.6"
    },
    "files": [
        "dist"
    ],
    "scripts": {
        "build": "rollup -c rollup.config.js"
    },
    "dependencies": {
        "@igloo-ui/tokens": "^2.1.0",
        "@hopper-ui/tokens": "^4.3.2",
        "classnames": "^2.3.2",
        "framer-motion": "^6.5.1",
        "react-aria": "^3.32.1",
        "react-stately": "^3.30.1"
    },
    "devDependencies": {
        "@igloo-ui/button": "^0.9.4"
    },
    "browserslist": [
        "> 1%",
        "last 2 versions",
        "IE >= 11"
    ]
}
