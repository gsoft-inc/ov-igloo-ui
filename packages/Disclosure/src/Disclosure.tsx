import * as React from 'react';
import cx from 'classnames';
import { useButton } from 'react-aria';
import { m, AnimatePresence, LazyMotion, domAnimation } from 'framer-motion';

import ChevronDown from '@igloo-ui/icons/dist/ChevronDown';

import './disclosure.scss';

export interface DisclosureProps extends React.ComponentProps<'div'> {
  /** The component content to display when expanding */
  children: React.ReactNode;
  /** Add a specific class to the disclosure */
  className?: string;
  /** Add a data-test tag for automated tests */
  dataTest?: string;
  /** The description of the disclosure */
  description?: React.ReactNode;
  /** An icon to show as the first element in disclosure */
  icon?: React.ReactNode;
  /** Whether or not the disclosure is expanded by default */
  isExpanded?: boolean;
  /** The title of the disclosure */
  title?: string;
}

const Disclosure: React.FunctionComponent<DisclosureProps> = (
  props: DisclosureProps
) => {
  const {
    children,
    className,
    dataTest,
    description,
    icon,
    isExpanded = false,
    title,
  } = props;

  const btnRef = React.useRef<HTMLButtonElement>(null);
  const [expanded, setExpanded] = React.useState(isExpanded);

  const { buttonProps } = useButton(
    {
      onPress: () => {
        setExpanded(!expanded);
      },
    },
    btnRef
  );

  const classes = cx('ids-disclosure', className);

  return (
    <div className={classes} data-test={dataTest}>
      <button
        {...buttonProps}
        ref={btnRef}
        className="ids-disclosure__header"
        aria-expanded={expanded}
      >
        {icon && <span className="ids-disclosure__header-icon">{icon}</span>}
        <span className="ids-disclosure__header-content">
          {title && (
            <span className="ids-disclosure__header-title">{title}</span>
          )}
          {description && (
            <span className="ids-disclosure__header-desc">{description}</span>
          )}
        </span>
        <ChevronDown size="medium" className="ids-disclosure__header-chevron" />
      </button>
      <LazyMotion features={domAnimation} strict>
        <AnimatePresence initial={false}>
          {expanded && (
            <m.div
              className="ids-disclosure__content"
              key="content"
              initial="collapsed"
              animate="open"
              exit="collapsed"
              variants={{
                open: {
                  opacity: 1,
                  height: 'auto',
                  overflow: 'hidden',
                  transitionEnd: {
                    overflow: 'visible',
                  },
                },
                collapsed: { opacity: 0, height: 0, overflow: 'hidden' },
              }}
              transition={{ duration: 0.5, ease: [0.04, 0.62, 0.23, 0.98] }}
            >
              {children}
            </m.div>
          )}
        </AnimatePresence>
      </LazyMotion>
    </div>
  );
};

export default Disclosure;
