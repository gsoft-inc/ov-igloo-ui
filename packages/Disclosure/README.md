# Disclosure

A disclosure is a widget that toggles the visibility of content by clicking a button.

<Example is="custom" />

<ReferenceLinks is="custom" />

## Installation

To install `@igloo-ui/disclosure` in your project, you will need to run the following command using [npm](https://www.npmjs.com/):

```bash
npm install @igloo-ui/disclosure
```

If you prefer [Yarn](https://classic.yarnpkg.com/en/), use the following command instead:

```bash
yarn add @igloo-ui/disclosure
```

## Usage

Then to use the component in your code just import it!

```jsx
import Substract from '@igloo-ui/icons/dist/Substract';

import Disclosure from '@igloo-ui/disclosure';

<Disclosure
  title="Diversity"
  icon={<Substract size="large" />}
  description="0.8pt in the last 30 days"
>
  <div style={{ background: '#F7F9FA', padding: '2.4rem' }}>
    Our organization values diversity.
  </div>
</Disclosure>;
```
