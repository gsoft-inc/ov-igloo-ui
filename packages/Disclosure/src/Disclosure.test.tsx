/**
 * @jest-environment jsdom
 */
import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import Substract from '@igloo-ui/icons/dist/Substract';

import Disclosure from './Disclosure';

const setup = (props = {}) => {
  return render(
    <Disclosure
      title="Diversity"
      icon={<Substract size="large" className="ids-disclosure__icon" />}
      description="0.8pt in the last 30 days"
      dataTest="ids-disclosure"
      {...props}
    >
      Our organization values diversity.
    </Disclosure>
  );
};

describe('Disclosure', () => {
  test('It should render without errors', () => {
    setup();
    const wrapper = screen.getByTestId('ids-disclosure');
    expect(wrapper).toBeInTheDocument();
  });

  test('It should render a snapshot', () => {
    const { asFragment } = setup();
    expect(asFragment()).toMatchSnapshot();
  });

  test('It should be expanded by default', () => {
    const { container } = setup({ isExpanded: true });
    const button = screen.getByRole('button');
    const content = container.querySelector('.ids-disclosure__content');
    expect(button).toHaveAttribute('aria-expanded', 'true');
    expect(content).toBeInTheDocument();
  });

  test('It should be expanded on button click', () => {
    const { container } = setup();
    const button = screen.getByRole('button');

    fireEvent.click(button);

    const content = container.querySelector('.ids-disclosure__content');

    expect(button).toHaveAttribute('aria-expanded', 'true');
    expect(content).toBeInTheDocument();
  });

  test('Check if props are displayed', () => {
    const { container } = setup({ isExpanded: true });
    const title = screen.getByText('Diversity');
    const desc = screen.getByText('0.8pt in the last 30 days');
    const icon = container.querySelector('.ids-disclosure__icon');
    expect(title).toBeInTheDocument();
    expect(desc).toBeInTheDocument();
    expect(icon).toBeInTheDocument();
  });
});
