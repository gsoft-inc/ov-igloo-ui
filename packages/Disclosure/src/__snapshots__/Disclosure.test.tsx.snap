// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Disclosure It should render a snapshot 1`] = `
<DocumentFragment>
  <div
    class="ids-disclosure"
    data-test="ids-disclosure"
  >
    <button
      aria-expanded="false"
      class="ids-disclosure__header"
      type="button"
    >
      <span
        class="ids-disclosure__header-icon"
      >
        <svg
          class="ids-icon ids-disclosure__icon"
          fill="none"
          height="32"
          viewBox="0 0 32 32"
          width="32"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="16"
            cy="16"
            r="10.667"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
          <path
            d="M10.343 16h11.314"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </span>
      <span
        class="ids-disclosure__header-content"
      >
        <span
          class="ids-disclosure__header-title"
        >
          Diversity
        </span>
        <span
          class="ids-disclosure__header-desc"
        >
          0.8pt in the last 30 days
        </span>
      </span>
      <svg
        class="ids-icon ids-disclosure__header-chevron"
        fill="none"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M5.47 8.47a.75.75 0 0 1 1.06 0L12 13.94l5.47-5.47a.75.75 0 1 1 1.06 1.06l-5.764 5.765a1.083 1.083 0 0 1-1.532 0L5.47 9.53a.75.75 0 0 1 0-1.06Z"
          fill="currentColor"
          fill-rule="evenodd"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
